src/core/border_router/routing_manager.cpp

```diff
@@ -458,20 +457,53 @@ void RoutingManager::EvaluateOmrPrefix(void)
     return;
 }
 
-void RoutingManager::PublishExternalRoute(const Ip6::Prefix &aPrefix, RoutePreference aRoutePreference, bool aNat64)
+void RoutingManager::EvaluatePublishingPrefix(const Ip6::Prefix &aPrefix)
 {
-    NetworkData::ExternalRouteConfig routeConfig;
+    // This method evaluates whether to publish or unpublish a given
+    // `aPrefix` as an external route in the Network Data. It makes a
+    // collective decision by checking with different sub-components to
+    // see whether or not each wants this prefix published and if so
+    // at what preference level and flags.
+    //
+    // Before calling this method, the sub-components need to make sure
+    // to update their internal state such that their `ShouldPublish()`
+    // provides the correct info.
 
-    OT_ASSERT(mIsRunning);
+    bool                             shouldPublish = false;
+    NetworkData::ExternalRouteConfig routeConfig;
 
     routeConfig.Clear();
     routeConfig.SetPrefix(aPrefix);
+    routeConfig.mPreference = NetworkData::kRoutePreferenceLow;
     routeConfig.mStable     = true;
-    routeConfig.mNat64      = aNat64;
-    routeConfig.mPreference = aRoutePreference;
 
-    SuccessOrAssert(
-        Get<NetworkData::Publisher>().PublishExternalRoute(routeConfig, NetworkData::Publisher::kFromRoutingManager));
+    // The order of checks is important. The Discovered Prefix Table is
+    // first followed by Local On Link Prefix manager and finally NAT64
+    // prefix manager.
+
+    shouldPublish = mDiscoveredPrefixTable.ShouldPublish(routeConfig);
+
+    if (mLocalOnLinkPrefix.ShouldPublish(routeConfig))
+    {
+        shouldPublish = true;
+    }
+
+#if OPENTHREAD_CONFIG_NAT64_BORDER_ROUTING_ENABLE
+    if (mNat64PrefixManager.ShouldPublish(routeConfig))
```

Comment:
Each component may independently update `routeConfig`. However, the last one to return will be the one that determines the preference. Is that what we want?

Reply:
Yes. Exactly. The order we check with each component is important.

We don't have this situation today, but we can have a component check previous preference value (set by some other component ahead of it in the list) and decide if it wants to change it or not.